pkg/discovery/kubernetes/control-point-discovery.go

```diff
@@ -0,0 +1,178 @@
+func (cpd *controlPointDiscovery) start() {
+	cpd.ctx, cpd.cancel = context.WithCancel(context.Background())
+
+	cpd.waitGroup.Add(1)
+
+	panichandler.Go(func() {
+		defer cpd.waitGroup.Done()
+
+		operation := func() error {
+			// Proceed only if we are the leader
+			for {
+				if cpd.election.IsLeader() {
+					// Proceed
+					break
+				} else {
+					// Check again in 5 seconds
+					time.Sleep(5 * time.Second)
+				}
+			}
+
+			// Discover all resources with /scale subresource
+			_, apiResourceListList, err := cpd.discoveryClient.ServerGroupsAndResources()
+			if err != nil {
+				log.Error().Err(err).Msg("Unable to get API resource list")
+				return err
+			}
+			scalableResources := []string{}
+
+			for _, apiResourceList := range apiResourceListList {
+				for _, apiResource := range apiResourceList.APIResources {
+					if apiResource.Kind == "Scale" {
+						// Get its parent resource
+						parentResource := strings.TrimSuffix(apiResource.Name, "/scale")
+						scalableResources = append(scalableResources, parentResource)
+					}
+				}
+			}
+
+			groupVersionResourceSet := make(map[schema.GroupVersionResource]string)
+			log.Info().Msgf("Scalable resources: %v", scalableResources)
+			for _, apiResourceList := range apiResourceListList {
+				for _, apiResource := range apiResourceList.APIResources {
+					// Check if apiResource.Name belongs to scalableResources
+					if slices.Contains(scalableResources, apiResource.Name) {
+						groupVersion, parseErr := schema.ParseGroupVersion(apiResourceList.GroupVersion)
+						if parseErr != nil {
+							log.Error().Err(parseErr).Msg("Unable to parse group version")
+							return parseErr
+						}
+						groupVersionResource := schema.GroupVersionResource{
+							Group:    groupVersion.Group,
+							Version:  groupVersion.Version,
+							Resource: apiResource.Name,
+						}
+						// Add to groupVersionResourceSet
+						groupVersionResourceSet[groupVersionResource] = apiResource.Kind
+					}
+				}
+			}
+
+			// Track each scalable resource
+			for groupVersionResource, kind := range groupVersionResourceSet {
+				log.Info().Msgf("Starting watch for Group: %s, Version: %s, Resource: %s", groupVersionResource.Group, groupVersionResource.Version, groupVersionResource.Resource)
+				resourceInterface := cpd.dynamicClient.Resource(groupVersionResource)
+
+				// watch for changes
+				_, controller := cache.NewInformer(
+					&cache.ListWatch{
+						ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+							return resourceInterface.List(cpd.ctx, options)
+						},
+						WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+							return resourceInterface.Watch(cpd.ctx, options)
+						},
+					},
+					&unstructured.Unstructured{},
+					0,
+					cpd.createResourceEventHandlerFuncs(groupVersionResource, kind),
+				)
+
+				// start controller
+				panichandler.Go(func() {
+					controller.Run(cpd.ctx.Done())
+				})
+			}
+
+			<-cpd.ctx.Done()
```

Comment:
to avoid deadlocks, can we have this wait on the channel surrounded by select statement or use waitgroups here ?

Reply:
There's actually no need to wait here since resource watcher are running in separate goroutines. Removed this. Actually found a real concurrency issue right above this